Statusbar update with zoom and Timecode/Frames/NoText

davidlamhauge opened the pull request:
This PR adds a label on the statusbar, that shows (1) No text, (2) Frame number or (3) Time code, for the scene.
I've put a ToolButton in the statusbar as well, so you can change your info type on the fly. The preferred info type is saved in the preferences.
For Time code, I've only MM:SS:FF and not Hours. Since we can have maximum 9999 frames in a scene, and 9999 frames at 3 fps equals less than an hour (55 minutes), I reckon that Hours is not needed, but it can easily be added.
![David_DavidScreencast_20200812_17_44_19](https://user-images.githubusercontent.com/1292931/90038037-34292080-dcc5-11ea-9040-9fe4dbd93e64.gif)

I realize that this solution differs somewhat from the feature suggested in 935, but if it is accepted it addresses #935

Jose-Moreno commented:
@davidlamhauge Nice job David! Sorry for not commenting much on your PR's and contributions lately, I've been very busy 😖 

On the matter of MM:SS:FF, I understand why you took out the hourly digits out, though I've seen programs that simply leave it with a single H digit as in H:MM:SS:FF when it's not sure to reach hourly long footage so e.g 10 seconds at 30FPS would be -> 0:00:10:30

Well I don't think it's a big deal anyway as long as people understand what is it they're looking at, and right now it's at least clear which system is being used 👍 

On a more technical level related to usability, would it be possible to get rid of the little dropdown box and make the label itself behave as the selection menu instead? I think clicking on the timecode itself would be nice. If that's not possible, I suppose changing the icon would help to make it more obvious that you can change the timecode format, otherwise it looks as if you can change the zoom% instead (which would also be nice! I'll ask for that later 😆 )

Other than that I think it will be a welcome addition. Thanks for looking out for all of these issues! 😊

davidlamhauge commented:
Thanks for the feedback, @Jose-Moreno !

About a digit for Hours, I'll wait and see what the other think. I don't think it's needed.

The tool button in the statusbar is there, because I was afraid that we would enter a 'How to change camera size'-problem, if it wasn't a noticable button. I do see, that if you choose 'No text', it will look like the button is about the zoom, so maybe there should be three slashes '---', periods '...' or something else, to indicate that there in fact is a label.

scribblemaniac commented:
I have not had a chance to test or look at the code yet, but I'm liking the way this looks. I did not even know you could add interactive widgets to the status bar.

I agree that hours aren't really necessary, however I would use the format `MM:SS.FF` (note the period instead of the last colon). This is how ffmpeg, and Audacity do it. Kdenlive uses a comma or a semicolon (might be local specific too). All these solutions help helps distinguish `MM:SS:FF` from `HH:MM:SS`.

I would also remove the separator between the button and the frames/timecode text to help emphasize that the button is for that segment. I think the button might also be better on the right, since that mimics the location of arrows on a combo box, but that is OS-specific and there is something nice about the button not moving when you change settings, so probably just leave it unless everyone else things the right side is better.

davidlamhauge commented:
I've replaced the last colon with a period. It looks good.
There is no separator in the status bar. There is a zoom label and a timecode label. I've tried to remove the visible border, but haven't been able to find the solution yet.

Jose-Moreno commented:
@scribblemaniac @davidlamhauge Excuse me, but that shouldn't be changed, the semicolon is proper in this case, please read this quote from the wiki on SMPTE Timecode.

> _While non-drop timecode is displayed with colons separating the digit pairs—"HH:MM:SS:FF"—drop-frame is usually represented with a semicolon (;) or period (.) as the divider between all the digit pairs—"HH;MM;SS;FF", "HH.MM.SS.FF"—or just between the seconds and frames—"HH:MM:SS;FF" or "HH:MM:SS.FF_

Since Pencil2D is not dropping frames (we don't use floating point FPS) it should use the colon previous to the last commit. You also see this in programs like After Effects:

![image](https://user-images.githubusercontent.com/4912907/90318637-7174f480-def7-11ea-829c-63fd24244fdc.png)

davidlamhauge commented:
@Jose-Moreno I know that colon is the 'proper' way to separate in timecode. That's why I used colon in the first case.
I didn't know that period and semicolon was used for drop frame timecode.
After seeing period as a divider between seconds and frames, I kind of prefer that to colon. It is like it divides time-units and movie-unts, but at the end of the day, I don't care. I'll put period or colon - what ever you guys can agree upon.

J5lx commented:
Nice to see this implemented! I don’t have the time to have a thorough look at this right now, but based on the screenshot there’s one tiny detail that I would prefer to be changed (other than what’s been discussed so far), namely the alignment. The convention for status bars is to put permanent stuff on the right and temporary stuff or contents that change often on the left. Left-aligning the zoom level made sense previously because it was only a “temporary” message (which was simply very rarely overwritten), but now with the truly permanent widgets the right side would be a better fit. The reason I’m bringing this up is because I also got some status bar improvements in the works which make use of the left area (and also an interactive zoom widget, which I believe someone mentioned above).

Nice work David!

scribblemaniac commented:
Fair point about the drop frame rule, however I should point out that `MM:SS:FF` is not proper SMPTE either as it's missing the hour section, which is definitely unnecessary here. Personally I still think there should be different separator to distinguish time units from frame units.

J5lx commented on app/src/mainwindow2.cpp:
```diff
@@ -31,6 +31,9 @@ GNU General Public License for more details.
 #include <QDateTime>
+#include <QLabel>
+#include <QToolButton>
+#include <QPushButton>
```
Unused include.

J5lx reviewed (changes requested):
Time for a review! A few smaller issues, but nothing too serious. As a general note, I’d prefer something more descriptive like “timecode format” over “timecode text (level)” or “timecode label enum” in the code.

For the timecode format, I agree more with the use of a colon. I’m not sure if I’ve ever seen a dot used for that in the context of animation, so chances are our users too are going to encounter the colon notation elsewhere. I’d say it’s better to get used to that early – once you know the meaning, the dot notation isn’t much of an improvement anyway (IMO).

Lastly, in the original issue it was recommended to include both the SMPTE and the FFF (S:ff) timecode as options. Considering the latter’s use in Japanese animation, it seems worth adding that to me.

davidlamhauge commented:
I think I have addressed all issues now.
I didn't know the Japanese time code S:FF, and tried to look it up, but couldn't find anything. I can't see the logic in calling it FFF, so I've called it SFF right now, but it can easily be changed, if it indded is FFF.

Jose-Moreno commented:
@davidlamhauge I think it's fine with what you have. Perhaps that japanese counter shown before was overly specific to some studios so it's not a problem.

If you see Opentoonz Timecode Iwa Effect they only have absolute frames and SMPTE (which is widely used in Japan and the Americas along NTSC)

![image](https://user-images.githubusercontent.com/4912907/92334759-87895700-f056-11ea-8b67-700d03b78b23.png)

davidlamhauge commented:
I have included the Japanese S:FF mode, and moved the info to the timeline, so we have four options: No text, Frames, SMPTE and SFF.

J5lx reviewed (changes requested):
I’m still unhappy about the implementation of the format selection menu and the TimeControls layout can use some tweaks. The other comments are basically just about cleanup, nothing serious. Once those things are taken care of, this PR is good to go!

davidlamhauge commented on app/src/mainwindow2.cpp:
```diff
@@ -121,6 +122,9 @@ MainWindow2::MainWindow2(QWidget* parent) :
 
+    mZoomLabel = new QLabel("");
+    ui->statusbar->addWidget(mZoomLabel);
+
```
I could remove these lines, but then I must remove
- the cal to updateZoomLabel() in line 128
- the connect with viewChanged and updateZoomLabel, in line 133
- the function updateZoomLabel() in line 1523
- the declaration of updateZoomLabel(), in line 127 in mainwindow2.h
Then there will be no zoom label at all.
Should I do it?

J5lx replied:
What I meant was essentially to restore the state of the status bar from before this PR. But if it is too much trouble, feel free to leave it as is and I will simply rebase my status bar PR as necessary.

J5lx replied:
(Resubmitting this comment because of some GitHub weirdness)

What I meant was essentially to restore the state of the status bar from before this PR. But if it is too much trouble, feel free to leave it as is and I will simply rebase my status bar PR as necessary.

davidlamhauge commented on core_lib/src/interface/timecontrols.cpp:
```diff
@@ -209,6 +246,18 @@ void TimeControls::updatePlayState()
 
+void TimeControls::showTimecodeLabelMenu()
+{
+    QPoint globalPos = mTimecodeSelect->cursor().pos();
+
+    QMenu* menu = new QMenu();
+    menu->addAction(tr("No text"), this, &TimeControls::noTimecodeText, 0);
+    menu->addAction(tr("Frames"),  this, &TimeControls::onlyFramesText, 0);
+    menu->addAction(tr("SMPTE Timecode"), this, &TimeControls::SMPTE_text, 0);
+    menu->addAction(tr("SFF Timecode"), this, &TimeControls::SFF_text, 0);
+
+    menu->exec(globalPos);}
```
I don't know what you ask for. In fact I looked at the palette implementation, that is listed below.
To me it is (almost) a 1:1 copy. What is it I don't see?
```
void ColorPaletteWidget::showContextMenu(const QPoint& pos)
{
    QPoint globalPos = ui->colorListWidget->mapToGlobal(pos);

    QMenu* menu = new QMenu;
    connect(menu, &QMenu::triggered, menu, &QMenu::deleteLater);

    menu->addAction(tr("Add"), this, &ColorPaletteWidget::addItem, 0);
    menu->addAction(tr("Replace"),  this, &ColorPaletteWidget::replaceItem, 0);
    menu->addAction(tr("Remove"), this, &ColorPaletteWidget::removeItem, 0);

    menu->exec(globalPos);
}
```

J5lx replied:
What I mean is that, instead of connecting a custom showContextMenu slot and manually creating the menu there, you can simply add the actions directly to the QToolButton (mTimecodeSelect) itself. Then you can remove the showContextMenu method entirely and QToolButton will take care of everything that’s necessary to actually show the menu. I will submit suggestions to show what I mean. It probably would have been easier if I had simply done that in the first place, sorry.

davidlamhauge commented on core_lib/src/interface/timecontrols.cpp:
```diff
@@ -291,10 +340,74 @@ void TimeControls::updateSoundScrubIcon(bool soundScrubEnabled)
+        break;
+    case TimecodeTextLevel::NOTEXT:
+    default:
+        mTimecodeLabel->setText("");
```
What can NOTEXT be, if it's not an empty string?

J5lx replied:
Sorry, I should have been more explicit. What I (failed to) imply was that the label should be completely hidden from the toolbar when NOTEXT is selected, so that it doesn’t consume any space.

davidlamhauge replied:
Done, but it's hard to see that it gives more space to the right.

J5lx replied:
You mean it’s hard to see that something else can be made to appear there?

Personally, I don’t find that empty space to be more meaningful than no space at all. Actually, that empty space looks kinda weird to me, which is why I requested this change.

Keep in mind that the default is to show the frame number. That means that if the timecode format has been set to “No Text”, the user *chose* to hide that part of the UI. Moreover, since they hid it, they already know that this option is available. That’s my take on the matter at least.

This aside, hiding the label doesn’t actually seem to work right now. That said, I don’t see any issues with the code either. I’ll try to figure out what’s going on there.

J5lx replied:
Figured it out. `addWidget(mTimecodeLabel)` returns a QAction which must be used to control the visibility.

> **Note:** You should use QAction::setVisible() to change the visibility of the widget. Using QWidget::setVisible(), QWidget::show() and QWidget::hide() does not work.

davidlamhauge replied:
I've tried what you said @J5lx . Here's the code:
```
    case TimecodeTextLevel::NOTEXT:
    default:
        mNoTimecodeAction->setVisible(false);
        mTimecodeLabel->setVisible(false);
        break;
```
...and it doesn't work. It is mNoTimecodeAction you're referring to? 
I've tried both with and without the " mTimecodeLabel->setVisible(false);".
I'm sorry, but there must be something I misunderstand.
Is the easiest way out, that you do the necessary changes, and I merge them?

J5lx replied:
When you call `addWidget(mTimecodeLabel)` you get a QAction as return value. Right now, that return value is ignored because it isn’t needed for anything else. But it is the action that setVisible() needs to be called on. I’ll send a PR in a moment.

J5lx replied:
Check davidlamhauge#16

J5lx commented on core_lib/src/interface/timecontrols.cpp:
```diff
@@ -40,13 +41,40 @@ void TimeControls::initUI()
+    mFps = mFpsBox->value();
+    mTimecodeSelect = new QToolButton(this);
+    mTimecodeSelect->setIcon(QIcon(":app/icons/new/svg/more_options.svg"));
+    mTimecodeSelect->setPopupMode(QToolButton::InstantPopup);
```
```suggestion
    mTimecodeSelect->setPopupMode(QToolButton::InstantPopup);
    mTimecodeSelect->addAction(mNoTimecodeAction = new QAction(tr("No text"), this));
    mTimecodeSelect->addAction(mOnlyFramesAction = new QAction(tr("Frames"), this));
    mTimecodeSelect->addAction(mSmpteAction = new QAction(tr("SMPTE Timecode"), this));
    mTimecodeSelect->addAction(mSffAction = new QAction(tr("SFF Timecode"), this));
```

J5lx commented on core_lib/src/interface/timecontrols.cpp:
```diff
@@ -188,6 +221,10 @@ void TimeControls::makeConnections()
+
+    connect(mFpsBox, spinBoxValueChanged, this, &TimeControls::setFps);
+    connect(mEditor, &Editor::fpsChanged, this, &TimeControls::setFps);
+    connect(mTimecodeSelect, &QToolButton::clicked, this, &TimeControls::showTimecodeLabelMenu);
```
```suggestion
    connect(mNoTimecodeAction, &QAction::triggered, this, &TimeControls::noTimecodeText);
    connect(mOnlyFramesAction, &QAction::triggered, this, &TimeControls::onlyFramesText);
    connect(mSmpteAction, &QAction::triggered, this, &TimeControls::SMPTE_text);
    connect(mSffAction, &QAction::triggered, this, &TimeControls::SFF_text);
```

J5lx commented on core_lib/src/interface/timecontrols.cpp:
```diff
@@ -209,6 +246,19 @@ void TimeControls::updatePlayState()
     }
 }
 
+void TimeControls::showTimecodeLabelMenu()
+{
+    QPoint globalPos = mTimecodeSelect->cursor().pos();
+
+    QMenu* menu = new QMenu();
+    menu->addAction(tr("No text"), this, &TimeControls::noTimecodeText, 0);
+    menu->addAction(tr("Frames"),  this, &TimeControls::onlyFramesText, 0);
+    menu->addAction(tr("SMPTE Timecode"), this, &TimeControls::SMPTE_text, 0);
+    menu->addAction(tr("SFF Timecode"), this, &TimeControls::SFF_text, 0);
+
+    menu->exec(globalPos);
+}
+
```
```suggestion
```

J5lx commented on core_lib/src/interface/timecontrols.h:
```diff
@@ -47,6 +48,8 @@ class TimeControls : public QToolBar
     int getRangeLower() { return mPlaybackRangeCheckBox->isChecked() ? mLoopStartSpinBox->value() : -1; }
     int getRangeUpper() { return mPlaybackRangeCheckBox->isChecked() ? mLoopEndSpinBox->value() : -1; }
 
+    void showTimecodeLabelMenu();
+
```
```suggestion
```

J5lx commented on core_lib/src/interface/timecontrols.h:
```diff
@@ -81,6 +90,8 @@ public slots:
     QSpinBox*    mLoopStartSpinBox = nullptr;
     QSpinBox*    mLoopEndSpinBox = nullptr;
+    QToolButton* mTimecodeSelect = nullptr;
+    QLabel*      mTimecodeLabel = nullptr;
```
```suggestion
    QLabel*      mTimecodeLabel = nullptr;
    QAction*     mNoTimecodeAction = nullptr;
    QAction*     mOnlyFramesAction = nullptr;
    QAction*     mSmpteAction = nullptr;
    QAction*     mSffAction = nullptr;
```

J5lx reviewed (commented):
Here’s what I meant wrt the menu.

J5lx reviewed (approved):
LGTM, merging.

J5lx commented:
@Jose-Moreno @CandyFace Shall we treat this as a fix for #935 as suggested in the OP or leave that issue open until we have an actual on-canvas display?

CandyFace commented:
Hmm.. It's not quite solution I had in mind for #935 but it does solve the problem. Alright let's close it.

Jose-Moreno commented:
@J5lx I feel the same as CandyFace. I suppose we can close it then. I'll do it after organizing the Issue via the project board to be related to this PR then. Thanks to everyone involved in this PR.
